bokeh/charts/_chartobject.py

```diff
@@ -39,6 +39,604 @@
+
+        Returns:
+            self: the chart object being configured.
+        """
+        self._tools = tools
+        return self
+
+    def filename(self, filename):
+        """Set the file name of your chart.
+
+        If you pass True to this argument, it will use ``untitled`` as a filename.
+
+        Args:
+            filename (str or bool): the file name where your plot will be written.
+
+        Returns:
+            self: the chart object being configured.
+        """
+        self._filename = filename
+        return self
+
+    def server(self, server):
+        """Set the server name of your chart.
+
+        If you pass True to this argument, it will use ``untitled``
+        as the name in the server.
+
+        Args:
+            server (str or bool): the name of your plot in the server
+
+        Returns:
+            self: the chart object being configured.
+        """
+        self._server = server
+        return self
+
+    def notebook(self, notebook=True):
+        """Show your chart inside the IPython notebook.
+
+        Args:
+            notebook (bool, optional) : whether to output to the
+                IPython notebook (default: True).
+
+        Returns:
+            self: the chart object being configured.
+        """
+        self._notebook = notebook
+        return self
+
+    # TODO: make more chain methods
+
+    def check_attr(self):
+        """Check if any of the underscored attributes exists.
+
+        It checks if any of the chained method were used. If they were
+        not used, it assigns the parameters content by default.
+        """
+        if not hasattr(self, '_title'):
+            self._title = self.__title
+        if not hasattr(self, '_xlabel'):
+            self._xlabel = self.__xlabel
+        if not hasattr(self, '_ylabel'):
+            self._ylabel = self.__ylabel
+        if not hasattr(self, '_legend'):
+            self._legend = self.__legend
+        if not hasattr(self, '_xscale'):
+            self._xscale = self.__xscale
+        if not hasattr(self, '_yscale'):
+            self._yscale = self.__yscale
+        if not hasattr(self, '_width'):
+            self._width = self.__width
+        if not hasattr(self, '_height'):
+            self._height = self.__height
+        if not hasattr(self, '_tools'):
+            self._tools = self.__tools
+        if not hasattr(self, '_filename'):
+            self._filename = self.__filename
+        if not hasattr(self, '_server'):
+            self._server = self.__server
+        if not hasattr(self, '_notebook'):
+            self._notebook = self.__notebook
+        if not hasattr(self, '_facet'):
+            self._facet = self.__facet
+        if not hasattr(self, '_palette'):
+            self._palette = self.__palette
+        if not hasattr(self, '_xgrid'):
+            self._xgrid = self.__xgrid
+        if not hasattr(self, '_ygrid'):
+            self._ygrid = self.__ygrid
```

Comment:
Once things are really settled and we are happy with everything, it's probably worth a little metaprogramming to reduce all this boilerplate (including the methods above too)